### ISSUE TITLE
A segmentation fault occurs when a map is cleared

### ISSUE THREAD

**Issue by Gyoma** [JacksonAllan/CC#8]
I think the title describes the problem quite accurately. So, here is the minimal code example to reproduce the problem:

```
int main( void )
{
    // Declare a map with int keys and short elements.
    map( int, short ) our_map;
    init( &our_map );

    // Inserting elements.
    for( int i = 0; i < 10; ++i )
        if( !insert( &our_map, i, i + 1 ) )
        exit( 1 ); // Out of memory.

    for_each(&our_map, key, val)
        printf("%d : %d \n", *key, *val);

    for( int i = 0; i < 10; ++i )
        erase( &our_map, i);

    // UB. Usually segmentation fault 
    for_each(&our_map, key, val)
        printf("%d : %d \n", *key, *val);

    cleanup( &our_map );
}
```

Seems like the bug is in this function (in particular, because of 0): 

```
// Returns a pointer-iterator to the first element, or end if the map is empty.
static inline void *cc_map_first(
  void *cntr,
  size_t el_size,
  uint64_t layout
)
{
  for( size_t i = 0; i < cc_map_hdr( cntr )->cap; ++i )
    if( *cc_map_probelen( cntr, i, el_size, layout ) )
      return cc_map_el( cntr, i, el_size, layout );

  return cc_map_end( cntr, 0 /* Dummy */, layout );
}
```

If change the last statement to

```
return cc_map_end( cntr, el_size /*0*/ /* Dummy */, layout );
```

then it seems working

**Comment by JacksonAllan** [comment]
Thanks for bringing this bug to my attention and hunting down the source! Needless to say, it is a rather critical issue. I'll release a new version, including this fix, unit tests for this specific scenario, a general tidy-up of the in-code documentation, within the next few days.

**Comment by JacksonAllan** [comment]
This has now been fixed (v1.0.4).
